### ISSUE TITLE
With GnuGo as ending bot, "Failed to start the bot, can not make a move, trying to restart: illegal move"

### ISSUE THREAD

**Issue by pdg137** [online-go/gtp2ogs#415]
I heard that the bot [Fuego](https://online-go.com/player/58542/) is not good at ending the game. So I'm experimenting with a Fuego-powered bot that uses GnuGo as a "ending bot", using a configuration like this:

```js
{
  apikey: "(my key)",
  bot: {
    command: ["/path/to/fuego"]
  },
  ending_bot: {
    command: ["/path/to/gnugo", "--mode", "gtp"],
    moves_to_allow_before_checking_ratio: 0.1,
    send_chats: false
  },
  greeting: {
    en: "Hi! I am a bot powered by fuego-1.1, gtp2ogs-8.0.3, and nixos-23.11."
  },
  farewell: {
    en: "Thanks for the game!"
  },
}
```

(It's all built and configured with Nix using the recipe [here](https://github.com/pdg137/ogs-gnugo-bot/tree/fuego).)

It worked fine with either GnuGo or Fuego as the main bot and no ending bot, but with GnuGo as "ending bot" I get periodic messages about an "illegal move":

```
Feb 25 21:03:32   Game.ts:364          [game 16827] Releasing bot(s)
Feb 25 21:03:32   Game.ts:576          [game 16827] Playing c7 { x: 2, y: 2, text: 'c7', resign: false, pass: false }
Feb 25 21:03:35   Game.ts:249          [game 16827] Opponent played f2
Feb 25 21:03:35   Game.ts:441          [game 16827] Acquiring main bot instance
Feb 25 21:03:35   Game.ts:451          [bot  1:23246] [game 16827] Acquired bot instance
Feb 25 21:03:35   Game.ts:468          [game 16827] [game 16827] Acquiring ending bot: 13 moves played out of 9 necessary to begin consulting ending bot
Feb 25 21:03:35   Game.ts:478          [bot  2:23253] [game 16827] Acquired resign bot instance
Feb 25 21:03:35   Bot.ts:210           [bot  2:23253] ? illegal move
Feb 25 21:03:35   Bot.ts:213           [bot  2:23253] 

Feb 25 21:03:35   Game.ts:364          [game 16827] Releasing bot(s)
Feb 25 21:03:35 ! Game.ts:544          [game 16827] Failed to start the bot, can not make a move, trying to restart
Feb 25 21:03:35 ! Game.ts:545          [game 16827] illegal move
Feb 25 21:03:35 ! Game.ts:636          [game 16827] illegal move
Feb 25 21:03:35   Game.ts:552          [game 16827] Unable to react correctly - re-connect to trigger action based on game state.
Feb 25 21:03:36   Game.ts:136          [game 16827] gamedata     W   Feijoa  [9x9]    
Feb 25 21:03:36   Game.ts:441          [game 16827] Acquiring main bot instance
Feb 25 21:03:36   Game.ts:451          [bot  1:23246] [game 16827] Acquired bot instance
Feb 25 21:03:36   Game.ts:468          [game 16827] [game 16827] Acquiring ending bot: 13 moves played out of 9 necessary to begin consulting ending bot
Feb 25 21:03:36   Game.ts:478          [bot  2:23253] [game 16827] Acquired resign bot instance
Feb 25 21:03:36   Game.ts:506          [bot  1:23246] Generating move for game 16827
```

They show up in the chat too:

https://beta.online-go.com/game/16827

I don't know whether the "ending bot" feature is generally working or not, but in any case the illegal move messages are annoying. Is this a known problem or did I do something wrong?

**Comment by pdg137** [comment]
Here's a more detailed log with showboard: true, verbosity: 2.

First, both bots are asked to generate a move:

```
Feb 28 23:36:43   Bot.ts:733           [bot 3] >>> genmove white
Feb 28 23:36:43   Bot.ts:733           [bot 4] >>> genmove white
Feb 28 23:36:43   Bot.ts:164           [bot  3:26183] SgDefaultTimeControl: timeLeft=10.0/1 remaining=1.0 timeMove=10.0
Feb 28 23:36:43   Bot.ts:164           [bot  3:26183] GoUctPlayer: No tree to reuse found
Feb 28 23:36:43   Bot.ts:191           [bot 4] <<< = H6
Feb 28 23:36:43   Bot.ts:733           [bot 4] >>> showboard
Feb 28 23:36:43   Bot.ts:191           [bot 4] <<< = 
   A B C D E F G H J
 9 . . . . . . . . . 9
 8 . . . . . O X X . 8
 7 . . + . X O X . . 7
 6 . . . . O . O O . 6
 5 . . X O X X O . . 5
 4 . X . O X O . . . 4
 3 . X X O O . + . . 3
 2 . . . . . . . . . 2     WHITE (O) has captured 0 stones
 1 . . . . . . . . . 1     BLACK (X) has captured 0 stones
   A B C D E F G H J
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] [0] SgUctSearch: max time reached (early abort)
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Count          83426
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] GamesPlayed    83426
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Nodes          3727890
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Time           3.3
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] GameLen        74.0 dev=6.7 min=44.0 max=114.0
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] InTree         15.7 dev=5.2 min=0.0 max=40.0
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Aborted        0%
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Games/s        24940.9
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Value          0.83
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] Sequence       D4 G4 G3 D5 J7 A2 F6 E6 F7 F8 H6 H3 H4 H2 G2 G1 F1 H1 J1
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] TimeInitTree   0.00
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] TimeRootFilter 0.00
Feb 28 23:36:46   Bot.ts:164           [bot  3:26183] GoUctPlayer: doing a search if early pass is possible
Feb 28 23:36:51   Bot.ts:164           [bot  3:26183] 0:05 | 0.482 | 124652 | 14.6 | H4 D4 D5 F6 G4 G3 F7 H3 H6 H5 J7 A3 B5 E6 D6 *
Feb 28 23:36:52   Bot.ts:164           [bot  3:26183] [0] SgUctSearch: maximum tree size 6250000 reached
Feb 28 23:36:52   Bot.ts:164           [bot  3:26183] SgUctSearch: pruning nodes with count < 16.0 (at time 5.6)
Feb 28 23:36:52   Bot.ts:164           [bot  3:26183] SgUctSearch: pruned size: 3277905 (53%) time: 0.1
Feb 28 23:36:52   Bot.ts:164           [bot  3:26183] [1] SgUctSearch: move cannot change anymore
Feb 28 23:36:52   Bot.ts:164           [bot  3:26183] GoUctPlayer: no early pass possible (no win)
Feb 28 23:36:52   Bot.ts:191           [bot 3] <<< = D4
Feb 28 23:36:52   Bot.ts:733           [bot 3] >>> showboard
Feb 28 23:36:52   Bot.ts:191           [bot 3] <<< = 
  A B C D E F G H J 
9 . . . . . . . . . 9
8 . . . . . . X X . 8
7 . . + . X . X O . 7
6 . . . . . . O . . 6
5 . . X . X X O . . 5
4 . X O O X O . . . 4
3 . X X O O O + . . 3
2 . O O . . . . . . 2  
1 . . . . . . . . . 1  B to play
  A B C D E F G H J
Feb 28 23:36:52   Game.ts:364          [game 16860] Releasing bot(s)
Feb 28 23:36:52   Game.ts:576          [game 16860] Playing d4 { x: 3, y: 5, text: 'd4', resign: false, pass: false }
Feb 28 23:36:52   Game.ts:223          [game 16860] game/16860/move: { game_id: 16860, move_number: 22, move: [ 3, 5, 9875 ] }
Feb 28 23:36:54   Game.ts:223          [game 16860] game/16860/move: { game_id: 16860, move_number: 23, move: [ 3, 4, 1711 ] }
```

You can see that the bots have different ideas about the state of the game, in particular at D5.

So when I play D5:

```
Feb 28 23:36:54   Game.ts:249          [game 16860] Opponent played d5
Feb 28 23:36:54   Game.ts:441          [game 16860] Acquiring main bot instance
Feb 28 23:36:54   Game.ts:451          [bot  3:26183] [game 16860] Acquired bot instance
Feb 28 23:36:54   Bot.ts:733           [bot 3] >>> play black d5
Feb 28 23:36:54   Bot.ts:191           [bot 3] <<< =
Feb 28 23:36:54   Bot.ts:733           [bot 3] >>> showboard
Feb 28 23:36:54   Bot.ts:191           [bot 3] <<< = 
  A B C D E F G H J 
9 . . . . . . . . . 9
8 . . . . . . X X . 8
7 . . + . X . X O . 7
6 . . . . . . O . . 6
5 . . X X X X O . . 5
4 . X O O X O . . . 4
3 . X X O O O + . . 3
2 . O O . . . . . . 2  
1 . . . . . . . . . 1  W to play
  A B C D E F G H J
Feb 28 23:36:54   Game.ts:468          [game 16860] [game 16860] Acquiring ending bot: 23 moves played out of 9 necessary to begin consulting ending bot
Feb 28 23:36:54   Game.ts:478          [bot  4:26190] [game 16860] Acquired resign bot instance
Feb 28 23:36:54   Bot.ts:733           [bot 4] >>> play black d5
Feb 28 23:36:54   Bot.ts:191           [bot 4] <<< ? illegal move
Feb 28 23:36:54   Bot.ts:210           [bot  4:26190] ? illegal move
Feb 28 23:36:54   Bot.ts:213           [bot  4:26190] 

Feb 28 23:36:54   Game.ts:364          [game 16860] Releasing bot(s)
Feb 28 23:36:54 ! Game.ts:544          [game 16860] Failed to start the bot, can not make a move, trying to restart
Feb 28 23:36:54 ! Game.ts:545          [game 16860] illegal move
Feb 28 23:36:54 ! Game.ts:636          [game 16860] illegal move
```

...GnuGo (bot 4) thinks it's an illegal move.  GnuGo had actually suggested D5 earlier in the game, and retained that move in its concept of the board state.  Shouldn't it be getting synced up with the actual board after every move somehow?

**Comment by anoek** [comment]
Fixed in 79f4148

**Comment by pdg137** [comment]
Thank you! It does seem to be working on 8.0.4.
